include/rdma, prov/shm,efa,util: fi_peer API changes

aingerson opened the pull request:
Update srx flow to take in match attributes
Rename peer_rx_entry->size to msg_size and clarify usage in man pages
Update providers and implementation to match updates
Add FI_PEER as a primary capability which signals support for being used as a peer

j-xiong commented on man/man3/fi_peer.3:
```diff
@@ -454,7 +454,7 @@ struct fi_peer_rx_entry {
     struct fi_peer_srx *srx;
     fi_addr_t addr;
-    size_t size;
+    size_t msg_size;
```
Man page changes should be made to the markdown files. The actual man page is updated automatically.

aingerson replied:
Oops! I think VS Code opened the wrong file for me... Updated. Thanks for catching!

j-xiong commented:
The build check error is real.  The `size` to `msg_size` conversion is missing somewhere:
```
prov/shm/src/smr_progress.c: In function ‘smr_alloc_cmd_ctx’:
prov/shm/src/smr_progress.c:919:17: error: ‘struct fi_peer_rx_entry’ has no member named ‘size’
  919 |         rx_entry->size = cmd->msg.hdr.size;
      |                 ^~
make[1]: *** [Makefile:29128: prov/shm/src/src_libfabric_la-smr_progress.lo] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/runner/work/libfabric/libfabric'
make: *** [Makefile:8872: all] Error 2
Error: Process completed with exit code 2.
```

aingerson commented:
Updated

shijin-aws commented:
AWS CI failure: failed to compile with ompi

```

*** Fault tolerance
checking if want fault tolerance... Enabled mpi (Specified auto)
checking if want checkpoint/restart enabled debugging option... Disabled

============================================================================
== Modular Component Architecture (MCA) setup
============================================================================
checking for subdir args...  '--prefix=/home/ec2-user/PortaFiducia/build/libraries/openmpi/v5.0.x/install' '--with-sge' '--without-verbs' '--with-libfabric=/home/ec2-user/PortaFiducia/build/libraries/libfabric/pr10314/install/libfabric' '--disable-man-pages' '--enable-ipv6' 'LDFLAGS=-Wl,--as-needed' '--enable-prte-prefix-by-default' '--enable-mca-dso=all' '--with-cuda=/usr/local/cuda' '--with-cuda-libdir=/usr/local/cuda/lib64/stubs' '--with-xpmem=/home/ec2-user/PortaFiducia/build/libraries/xpmem/install' '--with-xpmem-libdir=/home/ec2-user/PortaFiducia/build/libraries/xpmem/install/lib'
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking which components should be disabled... 
checking which components should be direct-linked into the library... 
checking which components should be run-time loadable... all 
checking which components should be static... default
checking for projects containing MCA frameworks... opal, ompi, oshmem

*** Configuring MCA for opal
checking for frameworks for opal... common, accelerator, allocator, backtrace, btl, dl, hwloc, if, installdirs, memchecker, memcpy, memory, mpool, patcher, pmix, rcache, reachable, shmem, smsc, threads, timer

+++ Configuring MCA framework common
checking for no configure components in framework common... 
checking for m4 configure components in framework common... ofi, sm, ucx

--- MCA component common:ofi (m4 configuration macro)
checking for MCA component common:ofi compile mode... dso
checking for ofi pkg-config name... /home/ec2-user/PortaFiducia/build/libraries/libfabric/pr10314/install/libfabric/lib/pkgconfig/libfabric.pc
checking if ofi pkg-config module exists... no
checking for ofi header at /home/ec2-user/PortaFiducia/build/libraries/libfabric/pr10314/install/libfabric/include... not found

Traceback (most recent call last):
  File "/home/ec2-user/PortaFiducia/libraries/openmpi/install.py", line 295, in <module>
    main()
  File "/home/ec2-user/PortaFiducia/libraries/openmpi/install.py", line 291, in main
```

wenduwan commented:
bot:aws:retest

shefty commented on man/fi_peer.3.md:
```diff
@@ -494,9 +502,17 @@ These calls are invoked by the peer provider to obtain the receive buffer(s)
+the posted receive passed to the peer. The peer provider is responsible for
+checking that an incoming message fits within the provided buffer space and
+generating truncation errors.
+The tag parameter is used for tagged messages.
```
I would consider documenting that tag must be set to 0 for non-tagged messages, so it's not left uninitialized.

aingerson replied:
Good call. Will update.

shefty commented on man/fi_getinfo.3.md:
```diff
@@ -338,6 +338,11 @@ additional optimizations.
   allow an initiator to target (or name) a specific receive context as
   part of a data transfer operation.
 
+*FI_PEER*
```
Are there changes which use this or needed updates to fabtests or the common flow?

aingerson replied:
What do you mean?

shefty replied:
This updates the man page only.  There's no code updated anywhere to use the new definition.

aingerson replied:
Ah gotcha. Yes this is all targeted for the link provider as a way to query providers that it is able to link together. But other providers can use it as well if using a peer provider directly (like efa should TECHNICALLY add FI_PEER to its shm fi_getinfo call). I can add that.

shefty replied:
It's okay if the link provider is going to set this on input.  But what provider is going to respond to the query?  How does setting this not just fail the call?

aingerson replied:
I don't follow your question - the link provider needs to make sure that all providers it's going to link together support the peer API so it queries the provider with caps->FI_PEER. shm will succeed. cxi will succeed. rxm will succeed. But for example psm3 probably won't. So the fi_getinfo call in the link layer will not include psm3 as a possible link. It will only include rxm+shm, cxi+shm, etc. If a user requests link with psm3+shm, it would return ENODATA instead of attempting to run it and getting an ENOSYS error. Are you saying you prefer link to fail when trying to open everything instead of not including the bad combination in the list of possible links? Or are you saying I missed adding this in shm (as a supported capability) and efa (as a requested shm capability)?

shefty replied:
E.g, in the shm provider:
```
#define SMR_TX_CAPS (OFI_TX_MSG_CAPS | FI_TAGGED | OFI_TX_RMA_CAPS | FI_ATOMICS)
#define SMR_RX_CAPS (FI_SOURCE | FI_RMA_EVENT | OFI_RX_MSG_CAPS | FI_TAGGED | \
		     OFI_RX_RMA_CAPS | FI_ATOMICS | FI_DIRECTED_RECV | \
		     FI_MULTI_RECV)

struct fi_info smr_info = {
	.caps = SMR_TX_CAPS | SMR_RX_CAPS | FI_MULTI_RECV | FI_LOCAL_COMM,
```

Where is FI_PEER being used as a cap anywhere?  I don't see any providers setting it.  It's not part of any common checks as a primary cap (e.g. included in ofi.h).  Fabtests getinfo_test doesn't include this.

aingerson replied:
Ok so option number 2, ie you're grumpy I didn't add it. Got it. Will add.

shijin-aws reviewed (approved):
efa provider change LGTM, thanks!

shijin-aws commented on prov/efa/src/rdm/efa_rdm_msg.c:
```diff
@@ -850,7 +854,6 @@ struct efa_rdm_ope *efa_rdm_msg_alloc_rxe_for_tagrtm(struct efa_rdm_ep *ep,
 		}
 		(*pkt_entry_ptr)->ope = rxe;
 
-		peer_rxe->size = efa_rdm_pke_get_rtm_msg_length(*pkt_entry_ptr);
```
I forget why I added this earlier... is it because tag matching also need size to support `fi_peek`? If so, is it safe to remove this?

shijin-aws replied:
https://github.com/ofiwg/libfabric/commit/d865260e0b2c0d6db7e487a2a4747fb0cdbd6d40

aingerson commented:
Updated with following changes:
- Added wording to man pages to indicate tag needs to be set to 0 for non-tagged case
- Specified FI_PEER as a domain capability since all the resources belong in the domain
- Added FI_PEER to ofi capabilities for filtering
- Added FI_PEER and FI_AV_USER_ID to fi_tostr
- Added FI_PEER to shm capabilities
- Added FI_PEER to shm getinfo query in efa
